pykafka/protocol.py

```diff
@@ -218,21 +236,51 @@ def pack_into(self, buff, offset):
         :param offset: The offset to start the write at
         """
         # NB a length of 0 means an empty string, whereas -1 means null
+        # Assuming a CreateTime timestamp, not a LogAppendTime.
         len_key = -1 if self.partition_key is None else len(self.partition_key)
         len_value = -1 if self.value is None else len(self.value)
-        fmt = '!BBi%dsi%ds' % (max(len_key, 0), max(len_value, 0))
-        args = (self.MAGIC,
+        protocol_version = self.protocol_version
+        # Only actually use protocol 1 if timestamp is defined.
+        if self.protocol_version == 1 and self.timestamp:
+            fmt = '!BBQi%dsi%ds' % (max(len_key, 0), max(len_value, 0))
+        else:
+            protocol_version = 0
+            fmt = '!BBi%dsi%ds' % (max(len_key, 0), max(len_value, 0))
+        args = [protocol_version,
                 self.compression_type,
                 len_key,
                 self.partition_key or b"",
                 len_value,
-                self.value or b"")
+                self.value or b""]
+        if protocol_version > 0:
+            args.insert(2, int(self.timestamp))
         struct.pack_into(fmt, buff, offset + 4, *args)
         fmt_size = struct.calcsize(fmt)
         data = buffer(buff[(offset + 4):(offset + 4 + fmt_size)])
         crc = crc32(data) & 0xffffffff
         struct.pack_into('!I', buff, offset, crc)
 
+    @property
+    def timestamp_d(self):
+        """Get the timestamp as a datetime, if valid"""
+        if self.timestamp not in (0, -1):
+            # Assuming a unix epoch
+            return datetime.utcfromtimestamp(self.timestamp / 1000.0)
+
+    @timestamp_d.setter
+    def timestamp_d(self, dt):
```

Comment:
I think this could use a slightly more descriptive name, maybe just `timestamp_dt` would be more widely understandable.